x-pack/plugins/security_solution/common/detection_engine/build_exceptions_filter.ts

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Filter } from '../../../../../src/plugins/data/common';
+import {
+  ExceptionListItemSchema,
+  CreateExceptionListItemSchema,
+  EntryMatch,
+  EntryMatchAny,
+  EntryNested,
+  entriesMatch,
+  entriesMatchAny,
+  entriesExists,
+  entriesNested,
+  EntryExists,
+} from '../../../lists/common';
+import { BooleanFilter, NestedFilter } from './types';
+import { hasLargeValueList } from './utils';
+
+type NonListEntry = EntryMatch | EntryMatchAny | EntryNested | EntryExists;
+interface ExceptionListItemNonLargeList extends ExceptionListItemSchema {
+  entries: NonListEntry[];
+}
+
+interface CreateExceptionListItemNonLargeList extends CreateExceptionListItemSchema {
+  entries: NonListEntry[];
+}
+
+export type ExceptionItemSansLargeValueLIsts =
+  | ExceptionListItemNonLargeList
+  | CreateExceptionListItemNonLargeList;
+
+export const chunkExceptions = (
+  exceptions: ExceptionItemSansLargeValueLIsts[],
+  chunkSize: number
+): ExceptionItemSansLargeValueLIsts[][] => {
+  if (exceptions.length === 0) {
+    return [];
+  } else if (exceptions.length <= chunkSize) {
+    return [exceptions];
+  } else {
+    const chunkedFilters: ExceptionItemSansLargeValueLIsts[][] = [];
+    for (let index = 0; index < exceptions.length; index += chunkSize) {
+      const exceptionsChunks = exceptions.slice(index, index + chunkSize);
+      chunkedFilters.push(exceptionsChunks);
+    }
+    return chunkedFilters;
+  }
+};
+
+export const buildExceptionItemFilter = (
+  exceptionItem: ExceptionItemSansLargeValueLIsts
+): BooleanFilter | NestedFilter => {
+  const { entries } = exceptionItem;
+
+  if (entries.length === 1) {
+    return createInnerAndClauses(entries[0]);
+  } else {
+    return {
+      bool: {
+        filter: entries.map((entry) => createInnerAndClauses(entry)),
+      },
+    };
+  }
+};
+
+export const createOrClauses = (
+  exceptionItems: ExceptionItemSansLargeValueLIsts[]
+): Array<BooleanFilter | NestedFilter> => {
+  return exceptionItems.map((exceptionItem) => buildExceptionItemFilter(exceptionItem));
+};
+
+export const buildExceptionFilter = ({
+  lists,
+  excludeExceptions,
+  chunkSize,
+}: {
+  lists: Array<ExceptionListItemSchema | CreateExceptionListItemSchema>;
+  excludeExceptions: boolean;
+  chunkSize: number;
+}): Filter | undefined => {
+  // Remove exception items with large value lists. These are evaluated
+  // elsewhere for the moment being.
+  const exceptionsWithoutLargeValueLists = lists.filter(
+    ({ entries }) => !hasLargeValueList(entries)
+  ) as ExceptionItemSansLargeValueLIsts[];
```

Comment:
I've seen a few other examples of using some new features of filter to allow the is guard to be placed like so:

```ts
  const exceptionsWithoutLargeValueLists = lists.filter(
    (item): item is ExceptionItemSansLargeValueLIsts => !hasLargeValueList(item.entries)
  );
```

And you will get a bit more type safety (possibly), but I don't think you can destructor.

Ref:
https://github.com/microsoft/TypeScript/issues/18562

Reply:
That did work, thank so much! Always feels nice to remove an as cast.